### ISSUE TITLE
Error reading sensor

### ISSUE THREAD

**Issue by julled** [krepost/dht22#1]
hi,

thank you for providing your driver!

i tried using it, it succesfully compiled on pi zero and i connected the sensor according to pinout.xyz to gpio23.

when loading the kernelmodule using

`insmod dht22.ko gpiopin=23`

i am getting
```

[ 1211.011137] dht22: hello, world
[ 1211.013277] dht22: GPIO pin 23
[ 1211.015087] dht22: interrupt 161
[ 1211.017147] dht22: device 240 created
[ 1214.049030] dht22: only detected 1 bits
[ 1214.050207] dht22: using gpio pin 23
[ 1214.051213] dht22: detected 1 edges
[ 1219.759523] dht22: device not ready
[ 1274.049025] dht22: only detected 1 bits
[ 1274.051210] dht22: using gpio pin 23
[ 1274.053318] dht22: detected 1 edges
[ 1284.969992] dht22: device not ready
[ 1288.615816] dht22: device not ready
[ 1334.049018] dht22: only detected 1 bits
[ 1334.051144] dht22: using gpio pin 23
[ 1334.053019] dht22: detected 1 edges

```

Using a python library with the same pin gives reasonable values. 

Do you have any idea?

**Comment by LarsE9n** [comment]
Try compiling it with debug information (remove the # from the line DEBUG=y in Makefile). This logs more data from the sensor reading.

I've had issues with the driver sometimes not being able to read data; unloading the module and then reloading it again a few seconds later usually fixed it. I never figured out what the underlying issue was. :-/

**Comment by julled** [comment]
This is already with debug info actiavted. It looks like it only detects "1 edge", so i thougth that maybe the pin initilaisiation is not working.

hm, i also tried to reload, but it didnt help.

**Comment by LarsE9n** [comment]
I see. There might be a timing issue somewhere. I'm not very happy with how I implemented the actual reading of the pulses. And I don't have access to an oscilloscope, so I was never able to figure out if the flakiness that I saw was due to the code or due to the sensor itself being flaky…

You mentioned that you used a python library successfully? Which library is that?

**Comment by julled** [comment]
> You mentioned that you used a python library successfully? Which library is that?

Adafruit_DHT22 lib (https://github.com/adafruit/Adafruit_Python_DHT) . But it says its deprecated, but nevertheless it works.

**Comment by LarsE9n** [comment]
Hi, I just wanted to mention that I tried the driver on a Pi Zero as well, and I get the same error as you. I have some ideas on how the driver could be more robust; will try to look into that over the next week or two.

**Comment by LarsE9n** [comment]
I used the wrong locking mechanism (mutex rather than spinlock) and also did a busywait which appears to have blocked interrupts on pi zero. I now rewrote the code to instead trigger a read once /dev/dht22 is opened by the user. This might lead to slightly more application errors (since a failed sensor read will trigger an error) but it simplifies the control flow. I have tested the new implementation on pi zero and pi 3.
